cpp/src/arrow/builder.cc

```diff
@@ -656,6 +658,206 @@ Status BooleanBuilder::Append(
+    const uint8_t* v = dict_builder_.GetValue(static_cast<int64_t>(index), &v_len);   \
+    return WrappedBinary(v, v_len);                                                   \
+  }                                                                                   \
+                                                                                      \
+  template <>                                                                         \
+  int DictionaryBuilder<Type, WrappedBinary>::HashValue(const WrappedBinary& value) { \
+    return HashUtil::Hash(value.ptr_, value.length_, 0);                              \
+  }                                                                                   \
+                                                                                      \
+  template <>                                                                         \
+  bool DictionaryBuilder<Type, WrappedBinary>::SlotDifferent(                         \
+      hash_slot_t index, const WrappedBinary& value) {                                \
+    int32_t other_length;                                                             \
+    const uint8_t* other_value =                                                      \
+        dict_builder_.GetValue(static_cast<int64_t>(index), &other_length);           \
+    return !(other_length == value.length_ &&                                         \
+             0 == memcmp(other_value, value.ptr_, value.length_));                    \
+  }                                                                                   \
+                                                                                      \
+  template <>                                                                         \
+  Status DictionaryBuilder<Type, WrappedBinary>::AppendDictionary(                    \
+      const WrappedBinary& value) {                                                   \
+    return dict_builder_.Append(value.ptr_, value.length_);                           \
+  }
+
+BINARY_DICTIONARY_SPECIALIZATIONS(StringType);
+BINARY_DICTIONARY_SPECIALIZATIONS(BinaryType);
+
+template class DictionaryBuilder<UInt8Type>;
+template class DictionaryBuilder<UInt16Type>;
+template class DictionaryBuilder<UInt32Type>;
+template class DictionaryBuilder<UInt64Type>;
+template class DictionaryBuilder<Int8Type>;
+template class DictionaryBuilder<Int16Type>;
+template class DictionaryBuilder<Int32Type>;
+template class DictionaryBuilder<Int64Type>;
+template class DictionaryBuilder<Date32Type>;
+template class DictionaryBuilder<Date64Type>;
+template class DictionaryBuilder<Time32Type>;
+template class DictionaryBuilder<Time64Type>;
+template class DictionaryBuilder<TimestampType>;
+template class DictionaryBuilder<FloatType>;
+template class DictionaryBuilder<DoubleType>;
+// template class DictionaryBuilder<BinaryType, WrappedBinary>;
```

Comment:
Does this need to be uncommented? These templates may need `ARROW_TEMPLATE_EXPORT`